mino/minows/rpc.go

```diff
@@ -0,0 +1,422 @@
+	serde.Message, error) {
+	stream, err := r.openStream(ctx, dest, pathCall)
+	if err != nil {
+		return nil, xerrors.Errorf("could not open stream: %v", err)
+	}
+
+	dec := gob.NewEncoder(stream)
+	err = r.send(dec, req)
+	if err != nil {
+		return nil, xerrors.Errorf("could not send request: %v", err)
+	}
+
+	enc := gob.NewDecoder(stream)
+	_, reply, err := r.receive(enc)
+	if err != nil {
+		return nil, xerrors.Errorf("could not receive reply: %v", err)
+	}
+	return reply, nil
+}
+
+func (r rpc) openStream(ctx context.Context, dest address,
+	path string) (network.Stream, error) {
+	pid := protocol.ID(r.uri + path)
+	stream, err := r.mino.host.NewStream(ctx, dest.identity, pid)
+	if err != nil {
+		return nil, xerrors.Errorf("could not open stream: %v", err)
+	}
+
+	go func() {
+		<-ctx.Done()
+		err := stream.Reset()
+		if err != nil {
+			r.logger.Error().Err(err).Msg("could not reset stream")
+		}
+	}()
+
+	return stream, nil
+}
+
+func (r rpc) send(enc *gob.Encoder, msg serde.Message) error {
+	from := r.mino.myAddr.location.Bytes()
+
+	var payload []byte
+	if msg != nil {
+		bytes, err := msg.Serialize(r.context)
+		if err != nil {
+			return xerrors.Errorf("could not serialize message: %v", err)
+		}
+		payload = bytes
+	}
+
+	err := enc.Encode(&Packet{Source: from, Payload: payload})
+	if errors.Is(err, network.ErrReset) {
+		return err
+	}
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	return nil
+}
+
+func (r rpc) receive(dec *gob.Decoder) (ma.Multiaddr, serde.Message, error) {
+	var packet Packet
+	err := dec.Decode(&packet)
+	if errors.Is(err, network.ErrReset) {
+		return nil, nil, err
+	}
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not decode packet: %v", err)
+	}
+
+	from, err := ma.NewMultiaddrBytes(packet.Source)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("could not unmarshal address: %v",
+			packet.Source)
+	}
+
+	if packet.Payload == nil {
+		return from, nil, nil
+	}
+	msg, err := r.factory.Deserialize(r.context, packet.Payload)
+	if err != nil {
+		return from, nil, xerrors.Errorf(
+			"could not deserialize message: %v",
+			err)
+	}
+	return from, msg, nil
+}
+
+func (r rpc) createOrchestrator(ctx context.Context,
+	initiator host.Host, streams []network.Stream) (*orchestrator, error) {
+	participant := r.mino.GetAddress().(address).location
+	myAddr, err := newOrchestratorAddr(participant, initiator.ID())
+	if err != nil {
+		return nil, xerrors.Errorf("could not create address: %v", err)
+	}
+
+	encoders := make(map[peer.ID]*gob.Encoder)
+	for _, stream := range streams {
+		encoders[stream.Conn().RemotePeer()] = gob.NewEncoder(stream)
+	}
+
+	o := &orchestrator{
+		logger: r.logger.With().Stringer("mino", myAddr).
+			Stringer("orchestrator", xid.New()).Logger(),
+		myAddr: myAddr,
+		rpc:    r,
+		outs:   encoders,
+		in:     make(chan Packet, MaxUnreadAllowed),
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(streams))
+	for _, stream := range streams {
```

Comment:
so many levels of indentation here. could we conveniently make use of an appropriately named function ?

Reply:
OK, with 8 spaces for a tab it's quite wide. IntelliJ only has 4 spaces for a tab, so it doesn't look that bad. I did extract the endless for-loop, tell me if it's better.